knative-operator/pkg/controller/knativeserving/knativeserving_controller.go

```diff
@@ -112,9 +118,21 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	}
 
 	gvkToResource := map[schema.GroupVersionKind]client.Object{
-		consolev1.GroupVersion.WithKind("ConsoleCLIDownload"): &consolev1.ConsoleCLIDownload{},
-		routev1.GroupVersion.WithKind("Route"):                &routev1.Route{},
+		routev1.GroupVersion.WithKind("Route"): &routev1.Route{},
+	}
+
+	// If console is installed add the ccd resource watcher, otherwise remove it avoid manager exiting due to kind not found.
+	// If we install the console later, this pod needs to be restarted as dynamically adding a watcher won't help since Serving reconciliation may not happen.
```

Comment:
This needs doc. We can't add a watcher for the console crds as we don't own them. Also web console cannot be uninstalled once isntalled so this should be enough. Also we use pod restarting in other occasions like tls secrets for internal encryption so not unexpected.

Reply:
This is unfortunate, really. I think, we can't add watcher to Console CRDs because they don't exist.

I got other idea how this could be done. It should be possible to watch on [ClusterVersion](https://docs.openshift.com/container-platform/4.11/rest_api/config_apis/clusterversion-config-openshift-io-v1.html) resource (`name == 'version'`). We could wait until its `.status.capabilities.enabledCapabilities` contains the `Console` capability. If such capability just became enabled, we probably still should wait a little (simple `.Get()` and `time.Sleep(250)` loop) until `ConsoleCLIDownload` is defined by Console Operator to avoid race with it. It should take at most a couple of seconds. After that from our point of view the Console is installed, and we could go ahead to create `ConsoleCLIDownload` resources.

Reply:
I dont feel this is the way to go (I thought about it regarding crds as well) as the reconciliation code requires a namespaced object related to Serving not clusterversion.
We depend on something we dont own and we need to get that information at installation time or after installation from an external resource. I am wondering how the OCP operators which provide oc deal with that?

Reply:
Ok console operator deals with the oc binaries: https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/clidownloads/controller.go

Reply:
> This is unfortunate, really. I think, we can't add watcher to Console CRDs because they don't exist.
We still can do `Watch(&source.Kind{Type: &apisextensionv1.CustomResourceDefinition{}}` I will take a look to see if we can implement it as a watch or an additional watcher.